docs/proposals/kubelet-eviction.md

```diff
@@ -241,11 +265,102 @@ only has guaranteed pod(s) remaining, then the node must choose to evict a
+around log evictions. `kubelet` can delete logs of the oldest containers first.
+Since logs from the first and the most recent incarnation of a container is the most important for most applications,
+kubelet can try to preserve these logs and aggresively delete logs from other container incarnations.
+
+### Delete unused images
+
+`kubelet` performs image garbage collection based on thresholds today. It uses a high and a low watermark.
+Whenever disk usage exceeds the high watermark, it removes images until the low watermark is reached.
+`kubelet` employs a LRU policy when it comes to deleting images.
+
+If `image` and `root` filesystems are unified, then `kubelet` will attempt to delete images whenever the thresholds are met
+and there are no more old logs to be deleted.
+
+If the `image` filesystem is separate from the `root` filesystem, `kubelet` will track the usage of `image` filesystem,
+and delete images only when `image` filesystem usage exceeds.
+
+### Evict pods
+
+There is no ability to specify disk limits for pods/containers today.
+Disk is a best effort resource. If required, kubelet can evict pods one at a time.
+`kubelet` will follow the [Eviction Strategy](#eviction-strategy) mentioned above for making eviction decisions.
+
+If `root` filesystem has met eviction thresholds, and there is no dedicated `image` filesystem,
+`kubelet` will first attempt to delete old containers first, then delete images,
+and then eventually evict pods if required.
+
+If `root` filesystem has met eviction thresholds, and if there is a dedicated `image` filesystem,
+`kubelet` will attempt to evict pods directly.
+
+If the `image` filesystem has met eviction thresholds, `kubelet` will first attempt to delete old containers first,
+then delete images, and then eventually evict pods if required.
+
+## Minimum eviction thresholds
+
+In certain scenarios, eviction of pods could result in reclamation of small amount of resources. This can result in
+`kubelet` hitting eviction thresholds in repeated successions. In addition to that, eviction of resources like `disk`,
+ is time consuming.
+
+To mitigate these issues, `kubelet` will have a per-resource `minimum-threshold`. Whenever `kubelet` observes
+resource pressure, `kubelet` will attempt to reclaim at least `minimum-threshold` amount of resource.
+
+Following are the internal `minimum-thresholds` that `kubelet` will use:
+
+`minMemoryEvictionThreshold = 0`
+`minDiskEvictionThreshold = 500Mi`
+
+*These values will not be user configurable.*
+
+## Deprecation of existing features
+
+`kubelet` has been freeing up disk space on demand to keep the node stable. As part of this proposal,
+some of the existing features/flags around disk space retrieval will be deprecated in-favor of this proposal.
+
+Following are the specific flags that will be deprecated:
+`--image-gc-high-threshold`
+`--image-gc-low-threshold`
+`--maximum-dead-containers`
```

Comment:
I briefly mentioned separating logs from containers in this proposal. The future I was thinking is that of not requiring dead containers at all. We will extract logs from old containers and immediately remove dead containers. We cannot do this today because kubelet uses dead containers as tombstones to generate pod status. If we add a simple checkpointing mechanism, then these tombstones are no longer necessary.
So this proposal should not be considered implemented until we can completely remove the need to keep dead containers around.

We can have a policy around the logs eviction in the future. For example, keep the first and the most recent crash logs of a recent pod for at-least an hour.

WDYT? These flags can exist while we implement this proposal, but they should be deprecated eventually.